cmd/juju/application/utils/names.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package utils
+
+import (
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+)
+
+var alphaChar = regexp.MustCompile("[A-Z]")
+var tailNumberSuffix = regexp.MustCompile("-[0-9]+$")
+
+// ValidateApplicationName takes a name and attempts to validate the application
+// name, before returning a reason why it's not valid.
+func ValidateApplicationName(name string) error {
+	if !names.IsValidApplication(name) {
+		// If the application has uppercase characters, bail out and explain
+		// why.
+		if alphaChar.MatchString(name) {
+			return errors.Errorf("invalid application name %q, unexpected uppercase character", name)
+		}
+		// If the application ends up being suffixed by a number, then we want
+		// to mention it to users why.
+		if tailNumberSuffix.MatchString(name) {
+			return errors.Errorf("invalid application name %q, unexpected number(s) found after hyphen", name)
+		}
+
+		index := strings.IndexFunc(name, validApplicationNameRune)
+		if index < 0 {
+			return errors.Errorf("invalid application name %q", name)
+		}
+
+		// We have to ensure that we don't break up multi-rune characters, by
+		// just selecting the index. Instead look at a slice of runes and use
+		// the first one.
+		invalidRune := []rune(name[index:])[0]
+		return errors.Errorf("invalid application name %q, unexpected character %c", name, invalidRune)
+	}
+
+	return nil
+}
+
+// validApplicationNameRune works out if there is a valid application rune.
+func validApplicationNameRune(r rune) bool {
+	if (r >= 'a' && r <= 'z') || unicode.IsNumber(r) || r == '-' {
+		return false
+	}
+	return true
+}
```

Comment:
I'm a little bit torn on this.

On the one hand, I'd rather that IsValidApplication returned a boolean and an error string, so that you'd have access to the reason for the failure everywhere that we call this. The code at this level of abstraction thus becomes much more compact: check for the error and return it if we find it, rather than doing forensics on why the error might be occurring.

That approach would put the hand translation of the regular expression in names/application.go closer to the regular expression itself. I'd even be tempted to factor out the regular expression. I'm not sure that the cpu cycles that it saves are worth the levels of indirection in telling the user about failures.

On the other hand, changing the return of IsValidApplication requires touching a lot of code, in a lot of packages. I understand the temptation to just put the wrapper here.

I like the change, overall. I just have mixed feelings about a wrapper being appropriate, here.

Reply:
Changing `IsValidApplication` is a pure no-go. I might as well re-write juju. 
Instead, I could lift this into the names package, that way we can share why things may fail and hopefully it may become a pattern for other name tags in that repo.

Reply:
@SimonRichardson +1 to moving this wrapper to names. That makes sense, and i think abstracts things more cleanly.